### PR TITLE
Implement the compiler part of EEP 49

### PR DESCRIPTION
**Pull request by bjorng**
This commit implements the compiler part of [EEP 49](https://github.com/erlang/eep/blob/172e818dcc0038de8fa881d394a9e79a2ec038ac/eeps/eep-0049.md) (Value-Based Error
Handling Mechanisms). Since the OTP Technical Board decided that the
`begin` keyword should be not be used to begin this construct, this
commit uses `maybe`.

Here is one of the examples rewritten to the syntax that this
commit supports:

    maybe
        Foo = bar(),
        {ok, X} <~ f(Foo),
        [H|T] <~ g([1,2,3]),
        . . .
    else
        {error, Y} ->
            {ok, "default"};
        {ok, _Term} ->
            {error, "unexpected wrapper"}
    end

Notes:

* The `maybe` ... `else` ... `end` construct is only allowed to be
  used if the compiler option `{enable_feature,maybe}` has been given.
  If not given, `maybe` and `else` will be interpreted as atoms, and
  code that use those atoms can be compiled. As currently implemented,
  the `enable_feature` option will not be recognized if given in the
  source code for the module. This limitation will be lifted later
  (there is ongoing work on a more robust and general implementation
  of `enable_feature`).

* The current version of the EEP states that the initial thought when
  creating the EEP was to use the `maybe` keyword to start the
  construct. `maybe` was also a favorite among the members of the OTB.
  Note that the atom `maybe` is used in several hundred places in OTP,
  and ultimately they will have to be updated when the `maybe`
  construct becomes mandatory in some later OTP release.

* The match operator is currently `<~`. There were objections at the
  OTB meeting that the same operator that was used in list comprehensions
  was used with a different meaning (`<-` in a list comprehension
  extracts the next element from a list, while `<~` matches an entire
  term).

* Regarding the other possible match operators mentioned in the EEP,
  `=~` is used for regular expression matching in Elixir (and Perl),
  and we might want to use it for the same purpose in Erlang.
  The `<|` could be useful in the future if we want to implement
  Elixir's pipe operator; in Erlang we would probably want to use
  both `<|` and `|>` (see
  https://erlangforums.com/t/what-did-you-first-think-of-the-erlang-syntax/365/9?u=bjorng).

* Variable scoping is similar to `try` ... `catch` ... `end`. That is,
  variables bound between `try` and `else` are unsafe (must not be
  used) bewtween `else` and `end`, and after the `end` keyword;
  variables bound between `else` and `end` are unsafe after the `end`
  keyword.

### CODE REVIEW

**Comment by meox** [comment]
which is the difference from https://github.com/erlang/otp/pull/5216?

**Comment by meox** [comment]
Why introducing the `<~` operator when we can simply reuse `<-` operator?

**Comment by bjorng** [comment]
> which is the difference from #5216?

As @ferd wrote, #5216 is first proof of concept implementation of EEP-49. I used their implementation as inspiration, tweaking some implementation details and also addressing notes from the OTP Technical Board meeting.

> Why introducing the `<~` operator when we can simply reuse `<-` operator?

See the third and fourth bullet points in the description of this pull request.

File: lib/compiler/src/v3_core.erl [hunk 1]
```diff
@@ -662,6 +690,84 @@ expr({block,_,Es0}, St0) ->
...
+    {[V1,V2,FailVar],St1} = new_vars(3, St0),
+
+    %% Translate the body of the letrec.
+    Fail = {maybe_else_fail,1},
+    Lanno = lineno_anno(L, St1),
+    {Es1,St2} = maybe_match_exprs(Es0, #c_var{name=Fail}, St1),
+
+    %% Translate the 'else' clauses. Note that we must not put the clauses
+    %% as the top-level clauses in the fun, because all shawdowing variables
+    %% in a fun head will be renamed.
```
[the remark this answers is not in the document]
Comment (ferd) [reply lib/compiler/src/v3_core.erl:756]: that's what the test file proves, nice.

[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/v3_core.erl:756]: Yes, I originally got it wrong, trying to be too smart put the clauses in the fun head.

File: lib/compiler/test/maybe_SUITE.erl [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    x = maybe nomatch <~ id(x) else E1 -> E1 end,
+
+    whatever = maybe
+                   AlwaysMatching <~ id(whatever),
+                   AlwaysMatching
+               else
+                   E2 -> E2
+               end,
+
+    ok.
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/test/maybe_SUITE.erl:83]: > A few cases I think are not being explicitly checked

Thanks, I will add tests for those cases.

File: lib/compiler/src/compile.erl [hunk 3]
```diff
@@ -1020,12 +1020,24 @@ do_parse_module(DefEncoding, #compile{ifile=File,options=Opts,dir=Dir}=St) ->
                         false ->
                             1
                     end,
+
+    ResWordFun = case proplists:get_value(enable_feature, Opts, []) of
+                     maybe ->
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/compile.erl:1025]: My plan is to wait for that mechanism and then rebase my branch.

**Pull request closed without merging**